tsl/test/src/remote/node_killer.c

```diff
@@ -17,88 +17,142 @@
 #include "export.h"
 #include "remote/connection.h"
 #include "remote/dist_txn.h"
+
 #include "test_utils.h"
 #include "connection.h"
 
-typedef struct RemoteNodeKiller
+static RemoteNodeKiller rnk_event = {
+	.conn = NULL,
+	.pid = -1,
+	.killevent = DTXN_EVENT_ANY,
+};
+
+static void
+kill_backend(const DistTransactionEvent event, void *data)
 {
-	pid_t pid;
-	TSConnection *conn;
-} RemoteNodeKiller;
+	RemoteNodeKiller *rnk = data;
 
-static char *kill_event = NULL;
-static RemoteNodeKiller *rnk_event = NULL;
+	if (event == rnk->killevent || rnk->killevent == DTXN_EVENT_ANY)
+	{
+		elog(WARNING, "kill event: %s", remote_dist_txn_event_name(rnk->killevent));
+		remote_node_killer_kill(&rnk_event);
+		remote_dist_txn_set_event_handler(NULL);
+	}
+}
+
+static DistTransactionEventHandler eventhandler = {
+	.handler = kill_backend,
+	.data = &rnk_event,
+};
 
 TS_FUNCTION_INFO_V1(ts_remote_node_killer_set_event);
 
-RemoteNodeKiller *
-remote_node_killer_init(TSConnection *conn)
+void
+remote_node_killer_init(RemoteNodeKiller *rnk, const TSConnection *conn,
+						const DistTransactionEvent event)
 {
-	RemoteNodeKiller *rnk = palloc(sizeof(RemoteNodeKiller));
+	int remote_pid = remote_connecton_get_remote_pid(conn);
+
+	if (remote_pid == -1)
+		elog(ERROR, "could not get PID of remote backend process");
 
+	MemSet(rnk, 0, sizeof(*rnk));
 	rnk->conn = conn;
-	rnk->pid = remote_connecton_get_remote_pid(conn);
+	rnk->pid = remote_pid;
+	rnk->killevent = event;
+}
 
-	/* do not throw error here on pid = 0 to avoid recursive abort */
-	/* remote_connection_report_error(ERROR, res, conn, false, sql);  */
-	return rnk;
+static int
+kill_backends(pid_t pid, int sig)
+{
+#ifdef HAVE_SETSID
+	/* If PostgreSQL is compiled with setsid support, then the running
+	 * PostgreSQL processes are in a process group and the negative pid kills
+	 * all the other processes in the same group. */
+	return kill(-pid, SIGTERM);
+#else
+	return kill(pid, SIGTERM);
+#endif
```

Comment:
```suggestion
static int
kill_backends(pid_t pid, int sig)
{
#ifdef HAVE_SETSID
	/* If PostgreSQL is compiled with setsid support, then the running
	 * PostgreSQL processes are in a process group and the negative pid kills
	 * all the other processes in the same group. */
	int err = kill(-pid, SIGTERM);
#else
	int err = kill(pid, SIGTERM);
#endif
    return err == ESRCH ? 0 : err;
```

Reply:
The suggested code change is incorrect. The return value of kill is either 0 or -1. ESRCH is not a possible return value; you need to check `errno` for that. But I don't really understand why one would do that and return 0 (success) in that case.